File: src/lexer/location.cpp
#include "location.hpp"

string Location::to_string() const {
    return path + ":" + std::to_string(line) + ":" + std::to_string(column);
}

void Location::advance(char const c) {
    if (c == '\n') {
        line++;
        column = 1;
    } else {
        column++;
    }
}

void Location::advance(string const& s) {
    for (char c : s) {
        advance(c);
    }
}

std::ostream& operator<<(std::ostream& os, Location const& location) {
    os << location.to_string();
    return os;
}

File: src/lexer/location.rs
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Location {
    path: String,
    line: usize,
    column: usize,
}

impl Location {
    pub fn new(path: String, line: usize, column: usize) -> Self {
        Self {
            path,
            line,
            column,
        }
    }

    pub fn advance(&mut self, c: &char) {
        if *c == '\n' {
            self.line += 1;
            self.column = 1;
        } else {
            self.column += 1;
        }
    }
}

impl std::fmt::Display for Location {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> core::fmt::Result {
        write!(f, "<{}:{}:{}>", self.path, self.line, self.column)
    }
}
